Inline tokens

UziTech opened the pull request:
## Description

Right now the `Lexer` returns the block tokens and the `Parser` sends the text to the `InlineLexer` which sends the information to the `Renderer`. There is no way to get the inline tokens before sending them to the `Renderer`.

This PR changes the `Lexer` to return all tokens with block tokens that have tokens inside of them in a `tokens` property on the token and changes the `Parser` to act on all tokens instead of the text from the block tokens.

Each token that has inline or block tokens will have a `tokens` property that is an array of tokens.

The `list` token will have an `items` property which will be an array of list items. Each list item will have a `tokens` property with the tokens inside the item.

This PR also adds a `raw` property to the tokens so a user can get the markdown that created that token.

fixes #1626
fixes #361

## TODO:

- [x] Write unit tests for `Lexer`, and `Parser`
- [x] Make sure demo works
	- You can see the new tokens [here](https://markedjs-t4fq3vpiu.now.sh/demo/?outputType=lexer&version=pr%3A1627)
- [x] Update docs
- [x] Refactor to make benchmarks faster
	- About 5-10% faster than master
	- It looks like this sped the compiled es5 version (`lib/marked.js`) up even more.

## Contributor

- [x] Test(s) exist to ensure functionality and minimize regression.
- [x] If submitting new feature, it has been documented in the appropriate places.

## Committer

In most cases, this should be a different person than the contributor.

- [ ] Draft GitHub release notes have been updated.
- [ ] CI is green (no forced merge required).
- [ ] Merge PR

UziTech commented:
> Each token that has inline tokens will have a tokens property that is an array of inline tokens.

I wasn't sure about how to add the inline tokens to the tokens array. Some tokens (like lists and blockquotes) have other tokens inside of them already so they have `start` and `end` tokens (e.g. `list_start`, `list_end`) before and after the tokens inside of them. I figured we could go one of three ways:

1. Change all tokens that have inline tokens to have `start` and `end` tokens (e.g. change `header` tokens to be `header_start` and `header_end` with the inline tokens between them.)
2. Give all tokens that have inline tokens a `tokens` property that contains their inline tokens.
3. Keep block tokens that can have other block tokens inside of them with `start` and `end` tokens and have all inline tokens added to a `tokens` property.

This PR currently does `#3` but I am leaning toward `#2`. I think `#2` is the cleanest but it would create major breaking changes for any dependents currently using the tokens from `marked.lexer`.

UziTech commented:
One thing I noticed while writing tests for the InlineLexer is that the `mangle` options mangles autolinks (e.g. `<test@example.com>`) but not plain email in gfm (e.g. `test@example.com`) [demo](https://markedjs-elv5po5p6.now.sh/demo/?outputType=html&text=%3Ctest%40example.com%3E%0A%0Atest%40example.com&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)

So I fixed it. [demo](https://markedjs-elv5po5p6.now.sh/demo/?outputType=html&text=%3Ctest%40example.com%3E%0A%0Atest%40example.com&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=pr%3A1627)

UziTech commented:
I'm not sure why travis ci isn't reporting but the tests are passing. [travis ci log](https://travis-ci.org/github/markedjs/marked/builds/667894929)

styfle commented on .travis.yml:
```diff
@@ -1,9 +1,9 @@
-    - stage: security scan 🔐
+  # fast_finish: true
+  # allow_failures:
+  #   - stage: security scan 🔐
```
Are these lines supposed to be commented out?

UziTech replied:
they can be since the security scan has been commented out for a while.

styfle commented on docs/USING_PRO.md:
```diff
@@ -91,29 +91,36 @@ slugger.slug('foo-1') // foo-1-2
 <h2 id="parser">The parser</h2>
 
-The parser is...
+The parser takes tokens as input and calls the renderer functions that are accosiated with those tokens.
```
```suggestion
The parser takes tokens as input and calls the renderer functions that are assocciated with those tokens.
```

UziTech commented:
@styfle what do you think about https://github.com/markedjs/marked/pull/1627#issuecomment-604113411?

styfle commented:
Can you give an example of what the token output would look like for the 3 examples?

Also, our PR demo doesn't seem to work anymore

![image](https://user-images.githubusercontent.com/229881/78052273-e1ba1980-734c-11ea-8c3d-4dc0b40b767e.png)

UziTech commented:
You will have to use the demo at https://markedjs-6edbij359.now.sh/demo/ since I changed the demo in this PR

UziTech commented:
```markdown
# heading

> blockquote
```

Tokens could look like:

1. Change all tokens that have inline tokens to have start and end tokens (e.g. change header tokens to be header_start and header_end with the inline tokens between them.)
```js
{type:"heading_start", raw:"# heading\\n\\n", depth:"1", text:"heading"}
{type:"text", raw:"heading", text:"heading"}
{type:"heading_end"}
{type:"blockquote_start", raw:"> blockquote"}
{type:"paragraph_start", raw:"blockquote", text:"blockquote"}
{type:"text", raw:"blockquote", text:"blockquote"}
{type:"paragraph_end"}
{type:"blockquote_end"}
```

2. Give all tokens that have inline tokens a tokens property that contains their inline tokens.
```js
{type:"heading", raw:"# heading\\n\\n", depth:"1", text:"heading", tokens: [
  {type:"text", raw:"heading", text:"heading"}
]}
{type:"blockquote", raw:"> blockquote", tokens: [
	{type:"paragraph", raw:"blockquote", text:"blockquote", tokens: [
	  {type:"text", raw:"blockquote", text:"blockquote"}
	]}
]}
```

3. Keep block tokens that can have other block tokens inside of them with start and end tokens and have all inline tokens added to a tokens property.
```js
{type:"heading", raw:"# heading\\n\\n", depth:"1", text:"heading", tokens: [
  {type:"text", raw:"heading", text:"heading"}
]}
{type:"blockquote_start", raw:"> blockquote"}
{type:"paragraph", raw:"blockquote", text:"blockquote", tokens: [
  {type:"text", raw:"blockquote", text:"blockquote"}
]}
{type:"blockquote_end"}
```

This PR implements #3 but I think it might be better to do #2 but that would be a bigger breaking change, and possibly have some bigger issues for our dependents.

styfle commented:
Thanks that is much easier to visualize!

I like option 2 the best because then we don't need _start and _end tokens, the nested `.tokens` property implies this.

UziTech commented:
I agree. I'll update the code to work like option 2.

UziTech commented:
I have rebased and implemented option 2 from https://github.com/markedjs/marked/pull/1627#issuecomment-606753680

It didn't make sense to keep the `InlineLexer` so I just merged it into the `Lexer`

UziTech commented:
I renamed the lexer functions and reordered the parameters for the lexer so they are consistent.

@styfle I think this is ready to merge if you approve.

experimentalPrivateRepos commented:
Hey @UziTech, thank you for the great work.
I would like to use the updated Lexer in a personal project of mine.
Do you have any idea when this new version will be available on npm?

UziTech commented:
@experimentalPrivateRepos We have a few more changes to get into the next release. I would estimate it will be about a month before the next release.

experimentalPrivateRepos commented:
@UziTech Alright, thank you for the quick response!

upupming commented:
> I have rebased and implemented option 2 from [#1627 (comment)](https://github.com/markedjs/marked/pull/1627#issuecomment-606753680)
> 
> It didn't make sense to keep the `InlineLexer` so I just merged it into the `Lexer`

Hi, since the `InlineLexer` class has been deleted, can you delete the type definitions in `@types/marked` too, thanks!
